backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/UniqId.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin.backend.konan.serialization
+
+import org.jetbrains.kotlin.backend.konan.llvm.localHash
+import org.jetbrains.kotlin.descriptors.*
+import org.jetbrains.kotlin.ir.declarations.IrDeclaration
+import org.jetbrains.kotlin.serialization.deserialization.descriptors.*
+import org.jetbrains.kotlin.metadata.KonanIr
+import org.jetbrains.kotlin.metadata.konan.KonanProtoBuf
+
+// This is an abstract uniqIdIndex any serialized IR declarations gets.
+// It is either isLocal and then just gets and ordinary number within its module.
+// Or is visible across modules and then gets a hash of mangled name as its index.
+data class UniqId (
+    val index: Long,
+    val isLocal: Boolean
+)
+
+// isLocal=true in UniqId is good while we dealing with a single current module.
+// To disambiguate module local declarations of different modules we use UniqIdKey.
+// It has moduleDescriptor specified for isLocal=true uniqIds.
+data class UniqIdKey private constructor(val uniqId: UniqId, val moduleDescriptor: ModuleDescriptor?) {
+    constructor(moduleDescriptor: ModuleDescriptor?, uniqId: UniqId)
+            : this(uniqId, if (uniqId.isLocal) moduleDescriptor else null)
```

Comment:
Would it be correct to add an assertion that if `uniqId.isLocal`, then `moduleDescriptor != null`?

Reply:
I have just added the double bang:  `if (uniqId.isLocal) moduleDescriptor!!`